### ISSUE TITLE
Replace photon source slider with on/off button

### ISSUE THREAD

**Issue by arouinfar** [phetsims/molecules-and-light#295]
Discussed in the 3/3/20 description meeting.

Adjusting the photon emission rate does not serve any of the primary learning goals. In the original design of the sim, the slider posed some challenges with student flow which we addressed by changing the shape of the thumb to look more tactile (see #83). Now that the sim has slow motion, a slider for the photon emission rate is less necessary. At times, having both a slider and slow motion option can lead to poor UX (e.g. low photon emission rate & slow motion).

An alternative would be to replace the slider with a `BooleanRoundStickyToggleButton` to turn the photon source on/off. We have used this control to turn on light/particle sources in many other sims including [Bending Light](https://phet.colorado.edu/sims/html/bending-light/latest/bending-light_en.html), [Waves Intro](https://phet.colorado.edu/sims/html/waves-intro/latest/waves-intro_en.html), and [Rutherford Scattering](https://phet.colorado.edu/sims/html/rutherford-scattering/latest/rutherford-scattering_en.html). These buttons often attract the user's first or second click.
<img src="https://user-images.githubusercontent.com/8419308/75829828-717ba080-5d6c-11ea-833e-25fdb7462587.png" width=200> <img src="https://user-images.githubusercontent.com/8419308/75829855-82c4ad00-5d6c-11ea-8689-a1f6010fb769.png" width=200> <img src="https://user-images.githubusercontent.com/8419308/75829865-8e17d880-5d6c-11ea-8681-a44acb432862.png" width=150>

Since this simulation is a spin-off of the Photon Absorption tab in The Greenhouse Effect, we decided to consult with the team currently working on the HTML5 design. There was a general consensus that this would be a good change, and appropriate for both sims.

We discussed the color of the button some in design meeting and wanted to choose something bright and attractive that doesn't necessarily coincide with a photon color. We initially liked the light blue used in Rutherford Scattering, but in context I don't think the color is ideal. It doesn't have great contrast against the gray light source and there's already a lot of blue elsewhere in the sim.
![image](https://user-images.githubusercontent.com/8419308/75830749-dc2ddb80-5d6e-11ea-8148-32f11598feea.png)

Here are some mockups showing a green (from Waves Intro) `BooleanRoundStickyToggleButton` on each light source. The button 35px in diameter, and its coordinates are fixed so it does not shift position between the various light sources.
![image](https://user-images.githubusercontent.com/8419308/75830630-92dd8c00-5d6e-11ea-884a-c4d8c3ba0481.png)
![image](https://user-images.githubusercontent.com/8419308/75830657-996c0380-5d6e-11ea-9957-5dda3d2e96c2.png)
![image](https://user-images.githubusercontent.com/8419308/75830671-a0931180-5d6e-11ea-93e6-f8dcbfd5fa08.png)
![image](https://user-images.githubusercontent.com/8419308/75830683-a557c580-5d6e-11ea-8a22-66b407c4181d.png)

Assigning to @emily-phet and @terracoda for review before @jessegreenberg proceeds with implementation.

**Comment by terracoda** [comment]
Green button looks good to me.

**Comment by emily-phet** [comment]
Yes - green. The sim has so much blue, a blue button doesn't stand out enough, i think.

**Comment by arouinfar** [comment]
Thanks @terracoda @emily-phet.

I realized we haven't specified what the photon emission rate should be. It will likely take a bit of tuning, but I think 85% of the current max rate might be a good place to start @jessegreenberg.

**Comment by jessegreenberg** [comment]
Looks great! @arouinfar is there a chance that the slider and photon emission rate will be needed in greenhouse? If not, I will remove things from PhotonAbsorptionModel related to this. Otherwise I will leave them in for that sim.

**Comment by jessegreenberg** [comment]
There is a substantial amount of code that could potentially l be removed related to emission frequency if that is definitely getting removed from both MAL and greenhouse, making a list here to remember in case it can be totally removed.

- PhotonAbsorptionModel.emissionFrequencyProperty
- PhotonAbsorptionModel.setPhotonEmissionPeriod
- PhotonAbsorptionModel.getSingleTargetPeriodFromFrequency
- PhotonAbsorptionModel.getSingleTargetFrequencyFromPeriod
- PhotonEmitterNode code related to changing 'brighness' of the emitter
- A11y code describing different photon emission rates
- EmissionRateControlSliderNode

**Comment by jessegreenberg** [comment]
Button added in the above commit. Current implementation updates the 'on' frequency and 'off' frequency when pressed without changing anything else.  But if we think that control of the emission frequency can be fully removed we will want to change the implementation so that things in https://github.com/phetsims/molecules-and-light/issues/295#issuecomment-597165243 are removed and Properpties that were used for slider are removed from PhET-iO API.

**Comment by arouinfar** [comment]
Discussed in the 3/10/20 description meeting.

* Button color/size/placement looks good.
* @arouinfar will double check with the Greenhouse team before @jessegreenberg removes all slider-related code for good.
* @jessegreenberg will create 2 query parameters to fine-tune the photon emission: one for the photon emission rate and one for the slow motion scale factor.

**Comment by jessegreenberg** [comment]
@arouinfar @terracoda @Matthew-Moore240 I made two query parameters `?slowSpeed` and `?emissionFrequency`. These are both values that can go from 0 to 1. They are both proportions of existing frequency/speeds. So ?`slowSpeed=0.5` means that at "slow" speed, the sim will run half as fast as "normal" speed. and `?emissionFrequency=0.25` means that photons will be emitted at 1/4 frequency comparted to fastest possible frequency of published version of the sim.

**Comment by jbphet** [comment]
The MAL sound design team was using the button in context during today's review meeting, and we have a suggestion: There should be no, or at least a lot less, delay between pushing the button and the emission of the first photon.  The current behavior, which can be seen on https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.37/phet/molecules-and-light_en_phet.html, seems a bit sluggish, and is particularly problematic when the sim is paused, then the button is pressed, and then the sim is stepped.  We noticed it took around 30 steps in this case before anything happened, which is likely to frustrate users.

**Comment by arouinfar** [comment]
>  There should be no, or at least a lot less, delay between pushing the button and the emission of the first photon

Agreed @jbphet. Can you look into this @jessegreenberg?

I also noticed that the color of the light bulb still depends on the `emissionFrequency`. For example, this is what the IR source looks like when `emissionFrequency=0.25`:
<img src="https://user-images.githubusercontent.com/8419308/76588668-2ae51f00-64ad-11ea-8ce9-34c7937cc8f6.png" width=200>

I think we should show gray bulb when the source is off and a fully saturated bulb when the source is on, regardless of what we choose the `emissionFrequency` to be. The color mapping no longer seems necessary now that we've removed the slider.

I like [emissionFrequency=0.6&slowSpeed=0.5](http://phettest.colorado.edu/molecules-and-light/molecules-and-light_en.html?ea&brand=phet&emissionFrequency=0.6&slowSpeed=0.5). If the team prefers a higher `emissionFrequency`, we'll probably need a slower `slowSpeed`, and something like [?emissionFrequency=0.75&slowSpeed=0.33](http://phettest.colorado.edu/molecules-and-light/molecules-and-light_en.html?ea&brand=phet&emissionFrequency=0.75&slowSpeed=0.33) could work. It's hard for me to say how slow the `slowSpeed` really needs to be -- I'm happy so long as the details of the vibration/rotation are clearly visible. I'd be really curious to hear what @terracoda @BLFiedler or @Matthew-Moore240's thoughts on this too.

**Comment by jbphet** [comment]
IMPORTANT NOTE: When the slow speed is finalized, this issue should be assigned to me, since the slow motion speeds for the sounds will need to be adjusted to match.

**Comment by terracoda** [comment]
@Matthew-Moore240, I tested sim [emissionFrequency=0.6&slowSpeed=0.5](http://phettest.colorado.edu/molecules-and-light/molecules-and-light_en.html?ea&brand=phet&emissionFrequency=0.6&slowSpeed=0.5) on two VO Speech Speeds 60 and 45.

**With VO Speech set to 60 or above**
- the context responses are great at normal speed, slow, and step forward

**With VO Speech 45** which seems slow to me (not sure what default is) 
- the context responses at normal speed are fine since they are pretty brief. With Sim Speed set to slow, the context responses get out of synch. 

We could either trim the detailed "slow motion" context responses a little more, or we could slow down slow motion. Or we could do a little of both.

Personally, I like the sound of two speeds when set with this query parameter emissionFrequency=0.6&slowSpeed=0.5

**Comment by terracoda** [comment]
@Matthew-Moore240, could you have a listen. You can change the numbers in the query parameter. I agree with @arouinfar and @BLFiedler that 0.6 and 0.5 are good settings for normal and slow.

**Comment by terracoda** [comment]
I will open a new issue, where @Matthew-Moore240 and I can review and optimize the slow and step forward responses.

**Comment by terracoda** [comment]
@Matthew-Moore240 and @terracoda have new responses on the design doc to discuss with team on Tuesday.

**Comment by terracoda** [comment]
@jessegreenberg, instead of making a new issue, I made added a comment in #285.

**Comment by jbphet** [comment]
Over Slack, @emily-phet said the following to me with respect to [this dev version](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.40/phet/molecules-and-light_en_phet.html):

> I think the default tempo for the photons is too high - I think closer to 70 bpm pace would be better. The sim feels a bit frantic now on startup. Here's a metronome at 70bpm. https://www.youtube.com/watch?v=wo-wr_tRkCw

I'm not sure how that will compare to the numbers discussed above for the speed settings, but I thought I'd add it as a point of comparison.  If the ultimate rate is substantially different than 70 bpm, I think it would be good to run it by @emily-phet.

**Comment by BLFiedler** [comment]
The above numbers (emissionFrequency=0.6&slowSpeed=0.5) are a little faster. Coming out to about 96 BPM with two tests @ 10 sec each.

**Comment by arouinfar** [comment]
Counting the number of photons emitted in 15s, [emissionFrequency=0.6](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.40/phet/molecules-and-light_en_phet.html?emissionFrequency=0.6&slowSpeed=0.5) is coming out to ~90 BPM. Bumping it down to [emissionFrequency=0.5](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.40/phet/molecules-and-light_en_phet.html?emissionFrequency=0.5&slowSpeed=0.5) is ~72 BPM, which I find to be a bit on the slow side, especially when combined with `slowSpeed=0.5`. Perhaps [emissionFrequency=0.55](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.40/phet/molecules-and-light_en_phet.html?emissionFrequency=0.55&slowSpeed=0.5) could be a happy medium?

**Comment by terracoda** [comment]
@arouinfar, I think it feels pretty good at 0.5 as the base emission frequency, and as long as the photon fires close to immediately in both fast and slow cases, we might be fine.

I am willing to compromise with 0.55, so really either is fine.

**Comment by terracoda** [comment]
Pinging @jbphet, to see if you are waiting on the outcome of this issue to finalize the sound?

**Comment by terracoda** [comment]
I'll just add, because the sound can be constant, I don't think it hurts to be a little cautious or conservative with the emission frequency. With the slider users could start out slow. With the button, the normal rate is immediate. We do want that normal rate to be comfortable for learners.

**Comment by terracoda** [comment]
@Matthew-Moore240 and @wandita can you both have a listen to the rates @arouinfar posted in https://github.com/phetsims/molecules-and-light/issues/295#issuecomment-603938661

These links are dev links so they should work for you both.

**Comment by wandita** [comment]
Hi Taliesin 
Thanks!
Where is the link? If the link is buried in Slack it is difficult. Do you have a direct link to it? 
wanda

**Comment by terracoda** [comment]
@wandita, apologies. Here are the choices for the emission rate:
1. Approximate rate of 90 BPM [emissionFrequency=0.6](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.40/phet/molecules-and-light_en_phet.html?emissionFrequency=0.6&slowSpeed=0.5) 
2. Approximate rate of 72 BPM [emissionFrequency=0.5](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.40/phet/molecules-and-light_en_phet.html?emissionFrequency=0.5&slowSpeed=0.5) 
3. Something in between [emissionFrequency=0.55](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.40/phet/molecules-and-light_en_phet.html?emissionFrequency=0.55&slowSpeed=0.5)

In all cases, the slow speed would be half the rate in each of the options posted above.

**Comment by wandita** [comment]
Thanks Taliesin!
I will listen shortly!
wanda

**Comment by wandita** [comment]
Dear Taliesin
Thank you so much! 
Do we have a way  not to have to go through the “story telling” and all the functionalities to go directly to the sounds”?  (i.e  correct me if I am wrong…the interest is to hear the different rates …let me know)
In my browser he linksopen the Simulation and I have to through the whole simulation to hear the sounds of interest.   Do you have the sounds only as a separate link? 
All the best
Wanda

**Comment by terracoda** [comment]
@wandita, again, my apologies the dev version do not yet have a name on the light source button. It is just read out as "button" on focus.

I don't have access to pre-recorded emission frequency rates.

I suggest we have a quick zoom call or you can load each version in a browser tab, and then just press the tab key until you get to the unnamed "button" in each version seperately. The un-named button is the light source button. Activate the unnamed button and you will hear the emission rate.

Just send me an email if you prefer a zoom call.

**Comment by terracoda** [comment]
@wandita, also I can see f I can get a later dev version with a properly named button.

**Comment by terracoda** [comment]
@wandita, here are 3 accessible light source switches:
1. approximately 90 BMP or [emission Frequency 0.6](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.41/phet/molecules-and-light_en_phet.html?switch&emissionFrequency=0.6&slowSpeed=0.5)
2. approximately 72 BMP or [emission Frequency 0.5](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.41/phet/molecules-and-light_en_phet.html?switch&emissionFrequency=0.50&slowSpeed=0.5)
3. something in the middle or [emission Frequency 0.55](https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.41/phet/molecules-and-light_en_phet.html?switch&emissionFrequency=0.55&slowSpeed=0.5)

To skip the intro at the top just hit the tab key. The Infrared Light Source Switch should be the first thing you land on.

**Comment by jbphet** [comment]
@terracoda asked:

> Pinging @jbphet, to see if you are waiting on the outcome of this issue to finalize the sound?

Yes, @Ashton-Morris and I will need to update the vibration and rotation sounds to match whatever rate is decided upon.

**Comment by terracoda** [comment]
In Tuesday's design meeting we decided on emission rate of 0.5 and the 0.5 of O.5

**Comment by jessegreenberg** [comment]
Over slack I asked if the PHotonAbsorptionModel still needs an emissionFrequencyProperty, and @arouinfar said
>When we discussed the on/off switch in the greenhouse brainstorming meeting, the consensus was that we wouldn’t need a slider or variable emission rate. I don’t think the emissionFrequency would necessarily be the same between the two sims, though.

The model and view could potentially be simplified by removing this Property now.

**Comment by jessegreenberg** [comment]
OK, this is done. Assigning to @jbphet for the sound portion of this issue, and to @arouinfar to make sure all is well with the new emission rates and play speed.

**Comment by jbphet** [comment]
I've looked through the code, and now that we have finalized rates for rotation and vibration, here are the rates that the sounds need to match up with:

Action | Normal Rate | Slow Motion Rate
--- | --- | ---
rotation | 1.1 rotations/sec | 0.55 rotations/sec
vibration | 5 cycles/sec | 2.5 cycles/sec

@Ashton-Morris - I just looked at the sounds that we currently have in master, and it looks like the normal speed rotation sounds are modulated at about 1 cycle/sec.  The normal speed vibration sound is at 5 cycles/sec, so it's fine as is.  Based on this, here is what we need updates for (and all sounds need to be spatialized):

- [ ] normal speed clockwise rotation
- [ ] normal speed counterclockwise rotation
- [ ] slow motion speed clockwise rotation
- [ ] slow motion speed counterclockwise rotation
- [ ] slow motion vibration

If it's easier, feel free to provide normal-speed vibration too.

Please assign this back to me when the sounds are available and I'll integrate them into the sim.

**Comment by Ashton-Morris** [comment]
[Here are the files.](https://drive.google.com/open?id=1dII2m4KtMuZyxGShyF0rXf7ZSGUKOkq6) They should be accurate in timing but I couldn't get the sim to load on my phettest so I didn't have a chance to get a screen recording in order to check them.


- vibration-slow-motion.mp3
- vibration-slow-motion.wav
- rotation-clockwise-slow-motion.mp3
- rotation-clockwise-slow-motion.wav
- rotation-counterclockwise-slow-motion.mp3
- rotation-counterclockwise-slow-motion.wav
- rotation-counterclockwise.mp3
- rotation-counterclockwise.wav
- rotation-clockwise.mp3
- rotation-clockwise.wav

**Comment by jbphet** [comment]
Sounds are integrated.  I'll have the design team review and will update this issue based on the results.

**Comment by terracoda** [comment]
@jbphet, did we verify everything for this issue. Can this issue be closed?

**Comment by jbphet** [comment]
Yep.  Closing.

**Comment by jbphet** [comment]
I just showed this sim to someone who is not a PhET team member, and they initially thought that the photons from the light source were knocking previously absorbed photons out of the molecule.  When I watched the simulation for a while, there really are a lot of times when the absorbed photon leaves the molecule right when a photon emitted from the source is over the molecule, so I can see why someone could end up with this impression.  The slider allowed users to experiment with the photon emission timing to see if there is a relationship, the button does not.  

I discussed this with @emily-phet, and she didn't think it was a huge problem, but was okay with me spending a little time to see whether I could tweak the emission timing a bit so that this happens less often.

**Comment by jbphet** [comment]
An examination of the code showed that the average hold time is 0.9 seconds, and the emission frequency is 0.8 seconds, so yeah, there is a very high probability that a photon will be close to a molecule when an absorbed photon is re-emitted.  I've adjusted the range for this so that this situation no longer arises at all.  @arouinfar - please check this out on master and let us know if you think this change should be made.  You can compare it to the latest dev version to see the previous behavior.

**Comment by jbphet** [comment]
@jessegreenberg - there is some odd looking code in `PhotonAbsorptionModel`, and it made it difficult for me to figure out what the emission frequency really is.  The methods `getSingleTargetPeriodFromFrequency` and `getSingleTargetFrequencyFromPeriod` use a constant called `MIN_PHOTON_EMISSION_PERIOD_SINGLE_TARGET` to calculate a frequency from a period, but I couldn't figure out why.  I suspect it may be left over from when we used to have a slider in here.  I mean, why not just define the period and use 1/period to define the frequency?

**Comment by jbphet** [comment]
Oh, and it looks like the method `getSingleTargetFrequencyFromPeriod` may not be used at all anymore.

**Comment by jbphet** [comment]
@BLFiedler - can you please also comment on whether the change to the re-emission behavior is worthwhile?  The discussion starts with [this comment](https://github.com/phetsims/molecules-and-light/issues/295#issuecomment-620125719).

**Comment by BLFiedler** [comment]
Short answer: If it's easy, sure.

From interviews I've done on MAL, I don't think I've run into that exact case.  My impression is that it's more likely the case someone believes that a photon is deflected (or bounces off) the molecule rather than being absorbed and re-emitted. It's possible that scenario is also more likely if a photon passes by the molecule the same time a photon is emitted and a learner tracks the new photon as being the one that just approached. This isn't frequent and can be ruled out pretty quickly by watching a few interactions or using the slow motion.

**Comment by arouinfar** [comment]
@jbphet the behavior in master looks good to me. I would be fine with making the change permanent.

**Comment by jessegreenberg** [comment]
Thanks, Im not sure what was going on here, it was indeed confusing! Frequency is no longer used when determining the rate of photon emission, we just have a defined emission period now. Reassigning to @jbphet to review or if there is more for https://github.com/phetsims/molecules-and-light/issues/295#issuecomment-620167989

**Comment by jbphet** [comment]
Thanks @jessegreenberg - that looks better.

@arouinfar and @BLFiedler - Thanks for your input.  I'll leave the change in place.  We can revisit if anything arises based on QA testing or subsequent interviews.
